Issue 7555 - fixed importimage

soltysh opened the pull request:
Fixes #7555.

@smarterclayton I've made slight modifications in the image-import command to address problems from the issue. Still there are some others I've marked as TODO for now. ptal

kargakis commented on pkg/cmd/cli/cmd/importimage.go:
```diff
@@ -132,8 +132,14 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
+				// TODO soltysh: what about the case where specs are only on tags?
+				if len(stream.Spec.DockerImageRepository) == 0 {
+					return fmt.Errorf("flag --all is applicable only to images with .spec.dockerImageRepository defined")
+				} else {
```
drop else

soltysh replied:
Right.

smarterclayton commented on pkg/cmd/cli/cmd/importimage.go:
```diff
@@ -132,8 +132,13 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
 
 		if all {
 			// importing a whole repository
+			// TODO soltysh: write tests to cover all the possible usecases!!!
```
Add a test now that covers the old behavior and new.

smarterclayton commented on pkg/cmd/cli/cmd/importimage.go:
```diff
@@ -132,8 +132,13 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
-				from = target
+				// TODO soltysh: what about the case where specs are only on tags?
+				if len(stream.Spec.DockerImageRepository) == 0 {
+					return fmt.Errorf("flag --all is applicable only to images with .spec.dockerImageRepository defined")
```
`spec.dockerImageRepository`, no leading `.`

soltysh commented:
@smarterclayton I've addressed the dot comment. I don't feel comfortable with import image not being tested at all, so I agree with your comment. I'll work on creating extensive tests of the ImageStreamImport created in imageimport.go as a followup if you don't mind.

smarterclayton commented:
Even a test-cmd test would be fine now.  I'd be happy with 2 or 3 more "import-image" commands added to images.sh

soltysh commented:
@smarterclayton I've reworked the image-import command to follow Complete-Validate-Run scheme and split it into testable chunks (commit 1). The bugfix and tests are in 2nd commit. ptal

soltysh commented:
[test]

soltysh commented:
Fixed gofmt problem.

soltysh commented:
One clean run, one flake filled new one for that. @smarterclayton still waiting your signoff.

smarterclayton commented:
Will review on Monday

smarterclayton commented on pkg/cmd/cli/cmd/importimage.go:
```diff
@@ -39,78 +40,241 @@ func NewCmdImportImage(fullName string, f *clientcmd.Factory, out io.Writer) *co
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunImportImage(f, out, cmd, args)
-			kcmdutil.CheckErr(err)
+			if err := opts.Complete(f, args, out); err != nil {
```
Don't force these to be usage errors, return usage errors and call CheckErr directly.

soltysh replied:
Right, `Complete` doesn't deal with usage errors here, the only ones there are from getting default namespace or clients.

smarterclayton replied:
The pattern here should be cmdutil.CheckErr(opts.Complete()) and the
same for the others.  Validate should return usage errors as
necessary.

soltysh replied:
I see, I must have been looking at older commands before this new pattern. Will change.

smarterclayton commented:
Change looks ok, can we add a few more branch tests to hack/test-cmd.sh

soltysh commented:
@smarterclayton what kind of tests are you thinking about?

soltysh commented:
@smarterclayton I've fixed this one issue, I'll be working on the tests both unit and cmd as a followup. I'd like to add `--dry-run` you've mentioned as well.

soltysh commented:
Let's have this one merged if you don't mind.

smarterclayton commented:
I'd like additional test-cmd tests here before we merge this that work
before and after.

soltysh commented:
> I'd like additional test-cmd tests here before we merge this that work before and after.

1) Refactored code but without my fix, lines [297-301](https://github.com/soltysh/origin/blob/ca0eecbc2c55bd0aca53bc39bfc0ef1420742f5e/pkg/cmd/cli/cmd/importimage.go#L297-L301):

```
if o.All {
    // importing a whole repository
    if len(from) == 0 {
        from = o.Target
    }
```

Covers the usecase where an image stream is found and we want to import all tags, without from being passed: `oc import-image myis --all`. This sets from to be the name of the image stream passed as an argument, which looking down the code is then being passed to importer as:

```
kapi.ObjectReference{
    Kind: "DockerImage",
    Name: from,
},
```

Which will generally fail when trying to read image metadata from remote repository. The only use-case where this will work is when your image stream uses the name of the images provided by docker hub (eg. mysql, python, etc.)

2) Refactored code but without my fix, lines [345-349](https://github.com/soltysh/origin/blob/ca0eecbc2c55bd0aca53bc39bfc0ef1420742f5e/pkg/cmd/cli/cmd/importimage.go#L345-L356):

```
} else {
    // create a new tag
    if len(from) == 0 {
        from = o.Target
    }
```

Covers the usecase where we import single tag, from existing image stream, but we omit the tag name, assuming default `latest` will be applied: `oc import-image myis`. Since the tag is not explicitly specified in the image stream's spec we fail to find it and we fall in the above `else` case where we again set from to point to image stream name. From this definition we create a `latest` tag pointing
to:

```
kapi.ObjectReference{
    Kind: "DockerImage",
    Name: from,
},
```

Further more we set above `ObjectReference` in the last `else` as main image stream's
`From`, which is not necessarily correct either. 

@smarterclayton having said that I'm not sure there was correct behavior before, that I can actually address with tests. I'm ok with adding more tests and I'm currently working on, it but I don't see any before that was actually working, though maybe I'm missing something here. If so please point me to the before usecase you're talking about.

soltysh commented:
@smarterclayton I've addressed your comment and added more tests, ptal

soltysh commented:
Of course I forgot to update tests after changing `Validate` - fixed.

soltysh commented:
It looks like flake from https://github.com/openshift/origin/issues/6077

soltysh commented:
@smarterclayton comment addressed

soltysh commented:
@smarterclayton test with checking no of imported tags. For your question why I've changed from importing `mysql` to `docker.io/mysql` is to remove ambiguity for having two mysql, where one is pointing to our image stream and second to hub's image.

smarterclayton commented:
Lgtm [merge]

soltysh commented:
Test flake from #7662, merge flake looks like #6077.
re-[merge]
